Adding support for OpenAI instance

carlotta94c opened the pull request:
## Purpose
This PR adds OpenAI API support for users who don't have access to Azure OpenAI. Additions:
- Calling OpenAI endpoint if env variable OPENAI_API_TYPE is set to openai (default models used: text-davinci-003 and gpt-3.5-turbo)
- Making bicep configuration conditional (Azure OpenAI resource is created only if param openAiType is set to 'azure')
- Adding instruction to switch from an Azure OpenAI endpoint (default) to a non-azure OpenAI instance
* ...

## Does this introduce a breaking change?
<!-- Mark one with an "x". -->
```
[ ] Yes
[x] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[x] Documentation content changes
[ ] Other... Please describe:
```

## How to Test
*  Get the code

```
git clone https://github.com/carlotta94c/azure-search-openai-demo
cd azure-search-openai-demo
git checkout openai_support_PR
npm install
```

* Test the code

Test deployment and app running in the cloud:

1. Navigate to the main.bicep file in the infra folder
2. Look for the openAiType parameter at line 16
3. Change its value to openai (instead of azure)
4. Configure the following app settings at line 135-136:
* OPENAI_API_ORGANIZATION={Your OpenAI organization}
* OPENAI_API_KEY={Your OpenAI API key}
5. Run azd up

To configure the local environment to use an existing OpenAI instance and test the app locally:

Run azd env set OPENAI_API_TYPE openai
Run azd env set OPENAI_API_ORGANIZATION {Your OpenAI organization}
Run azd env set OPENAI_API_KEY {Your OpenAI API key}

## What to Check
Verify that the following are valid
* Running 'azd up' without applying the changes to bicep file and env variables described above will ensure the same behavior of provisioning and deployment as before this PR (using Azure OpenAI resource is the default). 

## Other Information
Completion is not supported by OpenAI 'gpt-3.5-turbo'. For non-azure OpenAI instance the /chat endpoint is used.

tonybaloney commented on app/backend/approaches/retrievethenread.py:
```diff
@@ -70,12 +71,22 @@ def run(self, q: str, overrides: dict[str, Any]) -> Any:
-            n=1, 
-            stop=["\n"])
+
+        if(self.openai_type == "azure"):
```
the parenthesis aren't needed in Python

tonybaloney commented on app/backend/approaches/retrievethenread.py:
```diff
@@ -70,12 +71,22 @@ def run(self, q: str, overrides: dict[str, Any]) -> Any:
+                max_tokens=1024, 
+                n=1, 
+                stop=["\n"])
+        else:
```
These two conditions look almost identical? Is there a way to reduce duplication in this code

tonybaloney commented on app/backend/approaches/readretrieveread.py:
```diff
@@ -90,7 +91,12 @@ def run(self, q: str, overrides: dict[str, Any]) -> Any:
             input_variables = ["input", "agent_scratchpad"])
-        llm = AzureOpenAI(deployment_name=self.openai_deployment, temperature=overrides.get("temperature") or 0.3, openai_api_key=openai.api_key)
+        
+        if(self.openai_type == "azure"):
```
The parenthesis aren't required

tonybaloney commented on app/backend/approaches/readretrieveread.py:
```diff
@@ -90,7 +91,12 @@ def run(self, q: str, overrides: dict[str, Any]) -> Any:
-        llm = AzureOpenAI(deployment_name=self.openai_deployment, temperature=overrides.get("temperature") or 0.3, openai_api_key=openai.api_key)
+        
+        if(self.openai_type == "azure"):
+            llm = AzureOpenAI(deployment_name=self.openai_deployment, temperature=overrides.get("temperature") or 0.3, openai_api_key=openai.api_key)
```
`.get()` should take a the default argument as the second parameter, so you can replace `.get("temperature") or 0.3` with `get("temperature", 0.3)` in all occurrences ( I see a few in this PR )

tonybaloney commented on app/backend/approaches/readretrieveread.py:
```diff
@@ -40,11 +40,12 @@ class ReadRetrieveReadApproach(Approach):
     CognitiveSearchToolDescription = "useful for searching the Microsoft employee benefits information such as healthcare plans, retirement plans, etc."
 
-    def __init__(self, search_client: SearchClient, openai_deployment: str, sourcepage_field: str, content_field: str):
+    def __init__(self, search_client: SearchClient, openai_deployment: str, openai_type, sourcepage_field: str, content_field: str):
```
please add the type annotation as you have in the other constructor

tonybaloney commented on app/backend/approaches/readdecomposeask.py:
```diff
@@ -68,7 +70,11 @@ def run(self, q: str, overrides: dict[str, Any]) -> Any:
         cb_manager = CallbackManager(handlers=[cb_handler])
 
-        llm = AzureOpenAI(deployment_name=self.openai_deployment, temperature=overrides.get("temperature") or 0.3, openai_api_key=openai.api_key)
+        if(self.openai_type == "azure"):
```
No parenthesis required

tonybaloney commented on app/backend/approaches/chatreadretrieveread.py:
```diff
@@ -97,20 +98,48 @@ def run(self, history: Sequence[dict[str, str]], overrides: dict[str, Any]) -> A
-            stop=["<|im_end|>", "<|im_start|>"])
-
-        return {"data_points": results, "answer": completion.choices[0].text, "thoughts": f"Searched for:<br>{q}<br><br>Prompt:<br>" + prompt.replace('\n', '<br>')}
+        if(self.openai_type == "azure"):
```
No parenthesis required

tonybaloney commented on app/backend/app.py:
```diff
@@ -105,7 +124,8 @@ def ask():
 @app.route("/chat", methods=["POST"])
 def chat():
-    ensure_openai_token()
+    if(OPENAI_API_TYPE == "azure"):
```
No parenthesis

tonybaloney commented on app/backend/app.py:
```diff
@@ -89,7 +106,9 @@ def content_file(path):
 @app.route("/ask", methods=["POST"])
 def ask():
-    ensure_openai_token()
+    if(OPENAI_API_TYPE == "azure"):
```
No parenthesis

tonybaloney commented on app/backend/app.py:
```diff
@@ -52,16 +61,24 @@
     credential=azure_credential)
 blob_container = blob_client.get_container_client(AZURE_STORAGE_CONTAINER)
 
+if(OPENAI_API_TYPE == "azure"):
```
No parenthesis

tonybaloney commented on README.md:
```diff
@@ -91,36 +95,68 @@ It will look like the following:
+
+1. Navigate to the [main.bicep](./infra/main.bicep) file in the infra folder
+2. Look for the `openAiType` parameter at line 16
+3. Change its value to `openai` (instead of `azure`)
```
AZD has parameters. We shouldn't require the user to manually change the bicep file. Alternatively, use the `azd set` command with environment variables

mattmsft replied:
Agreed. We may want to consider splitting it into a separate bicep module as well.

tonybaloney commented on infra/main.bicep:
```diff
@@ -105,19 +115,30 @@ module backend 'core/host/appservice.bicep' = {
+      OPENAI_API_TYPE: openAiType
+      OPENAI_GPT_MODEL: openAiGptModelName
+      OPENAI_CHATGPT_MODEL: openAiChatGptModelName
+      /*OPENAI_API_KEY: {your OpenAI key}
```
See comment about using parameters instead of editing bicep

tonybaloney commented on notebooks/chat-read-retrieve-read.ipynb:
```diff
@@ -31,13 +37,19 @@
-    "openai.api_type = \"azure\"\n",
-    "openai.api_base = f\"https://{AZURE_OPENAI_SERVICE}.openai.azure.com\"\n",
-    "openai.api_version = \"2022-12-01\"\n",
+    "if(OPENAI_API_TYPE == \"azure\"):\n",
```
Remove parenthesis

tonybaloney commented on notebooks/chat-read-retrieve-read.ipynb:
```diff
@@ -101,12 +113,22 @@
-    "        max_tokens=32,\n",
-    "        stop=[\"\\n\"])\n",
+    "    \n",
+    "    if(OPENAI_API_TYPE == \"azure\"):\n",
```
Remove parenthesis

tonybaloney commented on notebooks/chat-read-retrieve-read.ipynb:
```diff
@@ -127,12 +149,20 @@
-    "    temperature=0.7, \n",
-    "    max_tokens=1024,\n",
-    "    stop=[\"<|im_end|>\", \"<|im_start|>\"])\n",
+    "if(OPENAI_API_TYPE == \"azure\"):\n",
```
Remove parenthesis

tonybaloney commented:
Thanks for the PR @carlotta94c this looks like a really useful feature. I've added some nitpick comments in the review, but the most important one is the use of parameters to set the OpenAI type instead of editing bicep files.

carlotta94c commented:
> Thanks for the PR @carlotta94c this looks like a really useful feature. I've added some nitpick comments in the review, but the most important one is the use of parameters to set the OpenAI type instead of editing bicep files.

Hi @tonybaloney, huge thanks for your prompt review! I believe I've addressed the requested changes with my last commit.

tonybaloney commented on app/backend/app.py:
```diff
@@ -33,14 +38,18 @@
-openai.api_type = "azure"
-openai.api_base = f"https://{AZURE_OPENAI_SERVICE}.openai.azure.com"
-openai.api_version = "2022-12-01"
+if(OPENAI_API_TYPE == "azure"):
```
Ah sorry one still left

carlotta94c replied:
Good catch, just removed!

carlotta94c commented:
closing this outdated PR as I opened a new one https://github.com/Azure-Samples/azure-search-openai-demo/pull/507
